### PR TITLE
[Enhance] Add type hint in `focal_loss.py`.

### PR DESCRIPTION
**Pull request by WINDSKY45**
## Motivation

To get gift.

## Modification

Add type hint in focal_loss.py.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcv/ops/focal_loss.py [hunk 1]
```diff
@@ -122,12 +131,13 @@ def symbolic(g, input, target, gamma, alpha, weight, reduction):
...
     @staticmethod
     def forward(ctx,
-                input,
-                target,
+                input: torch.Tensor,
+                target: Union[torch.Tensor, torch.LongTensor,
+                              torch.cuda.LongTensor],
                 gamma=2.0,
                 alpha=0.25,
                 weight=None,
```
Comment (HAOCHENYE) [inline mmcv/ops/focal_loss.py:139]: missing Type Hints

Comment (WINDSKY45) [reply mmcv/ops/focal_loss.py:139]: I have fixed it.

File: mmcv/ops/focal_loss.py [hunk 2]
```diff
@@ -122,12 +131,13 @@ def symbolic(g, input, target, gamma, alpha, weight, reduction):
 
     @staticmethod
     def forward(ctx,
-                input,
-                target,
+                input: torch.Tensor,
+                target: Union[torch.Tensor, torch.LongTensor,
```
Comment (zhouzaida) [inline mmcv/ops/focal_loss.py:135]: target should be LongTensor.

Comment (WINDSKY45) [reply mmcv/ops/focal_loss.py]: I have fixed it.

File: mmcv/ops/focal_loss.py [hunk 3]
```diff
@@ -88,14 +92,18 @@ def backward(ctx, grad_output):
...
+                 weight: Optional[torch.Tensor] = None,
+                 reduction: str = 'mean') -> None:
         super().__init__()
         self.gamma = gamma
         self.alpha = alpha
         self.register_buffer('weight', weight)
         self.reduction = reduction
 
-    def forward(self, input, target):
+    def forward(self, input: torch.Tensor, target: torch.LongTensor):
```
Comment (zhouzaida) [inline mmcv/ops/focal_loss.py:106]:
```suggestion
    def forward(self, input: torch.Tensor, target: torch.LongTensor) -> torch.Tensor:
```

File: mmcv/ops/focal_loss.py [hunk 4]
```diff
@@ -194,14 +204,18 @@ def backward(ctx, grad_output):
...
```
Comment (zhouzaida) [inline mmcv/ops/focal_loss.py:218]:
```suggestion
    def forward(self, input: torch.Tensor, target: torch.LongTensor) -> torch.Tensor:
```

Comment (WINDSKY45) [reply mmcv/ops/focal_loss.py]: I have fixed it.

**Review by imabackstabber — APPROVED** [review]
LGTM

**Review by teamwong111 — APPROVED** [review]

File: mmcv/ops/focal_loss.py [hunk 5]
```diff
@@ -27,12 +31,12 @@ def symbolic(g, input, target, gamma, alpha, weight, reduction):
...
     @staticmethod
     def forward(ctx,
-                input,
-                target,
-                gamma=2.0,
-                alpha=0.25,
-                weight=None,
-                reduction='mean'):
+                input: torch.Tensor,
+                target: torch.LongTensor,
```
Comment (zhouzaida) [inline mmcv/ops/focal_loss.py:35]: Should target be Union[torch.LongTensor, torch.cuda.LongTensor]?

File: mmcv/ops/focal_loss.py [hunk 6]
```diff
@@ -122,12 +133,12 @@ def symbolic(g, input, target, gamma, alpha, weight, reduction):
...
     def forward(ctx,
-                input,
-                target,
+                input: torch.Tensor,
+                target: torch.LongTensor,
                 gamma=2.0,
                 alpha=0.25,
                 weight=None,
-                reduction='mean'):
+                reduction='mean') -> torch.Tensor:
```
Comment (zhouzaida) [inline mmcv/ops/focal_loss.py:141]: missing type hints for those arguments

File: mmcv/ops/focal_loss.py [hunk 7]
```diff
@@ -88,14 +92,21 @@ def backward(ctx, grad_output):
...
         self.gamma = gamma
         self.alpha = alpha
         self.register_buffer('weight', weight)
         self.reduction = reduction
 
-    def forward(self, input, target):
+    def forward(
+        self, input: torch.Tensor,
+        target: Union[torch.LongTensor,
+                      torch.cuda.LongTensor]) -> torch.Tensor:
```
Comment (zhouzaida) [inline mmcv/ops/focal_loss.py:109]:
```suggestion
    def forward(
        self, input: torch.Tensor,
        target: Union[torch.LongTensor, torch.cuda.LongTensor],
    ) -> torch.Tensor:
```

**Review by zhouzaida — APPROVED** [review]

File: mmcv/ops/focal_loss.py [hunk 8]
```diff
@@ -88,14 +92,21 @@ def backward(ctx, grad_output):
...
         super().__init__()
         self.gamma = gamma
         self.alpha = alpha
         self.register_buffer('weight', weight)
         self.reduction = reduction
 
-    def forward(self, input, target):
+    def forward(
+        self, input: torch.Tensor,
+        target: Union[torch.LongTensor, torch.cuda.LongTensor],
```
Comment (zhouzaida) [inline mmcv/ops/focal_loss.py:108]:
```suggestion
        self,
        input: torch.Tensor,
        target: Union[torch.LongTensor, torch.cuda.LongTensor],
```

**Review by HAOCHENYE — APPROVED** [review]

**Pull request merged**